connector/src/test/scala/com/basho/riak/spark/rdd/connector/RiakConnectorTests.scala

```diff
@@ -32,14 +46,136 @@ class RiakConnectorTests {
+
+    // check that all three sessions contains different client instances
+    assertNotSame(firstSession.unwrap(), secondSession.unwrap())
+    assertNotSame(firstSession.unwrap(), thirdSession.unwrap())
+
+    // this session should be different from first due to eviction despite the same config
+    val fourthSession = connector.openSession()
+    assertNotSame(firstSession.unwrap(), fourthSession.unwrap())
+  }
+
+  @Test
+  def evictedSessionWithNoReferencesShouldBeClosedImmediately(): Unit = {
+    // create session
+    val session = connector.openSession()
+    val clientMock = session.unwrap()
+    session.close()
+    assertTrue(session.isClosed)
+
+    // check that first session is still alive
+    verify(clientMock, never()).shutdown()
+
+    // check that first session is still in cache (the same RiakClient instance will be returned)
+    val sameClientSession = connector.openSession()
+    assertSame(clientMock, sameClientSession.unwrap())
+    sameClientSession.close()
+
+    // evict session (second opening is required due to cache size == 2)
+    val newClientMock = connector.openSession(Some(HostAndPort.hostsFromString("tmp1", 0).toSeq)).unwrap()
+    connector.openSession(Some(HostAndPort.hostsFromString("tmp2", 0).toSeq))
+
+    // check that first session was evicted and closed
+    assertNotSame(clientMock, connector.openSession().unwrap())
+    verify(clientMock, only()).shutdown()
+
+    // check that second session was not closed (method "java.lang.AutoCloseable.close" was not called)
+    verify(newClientMock, never()).shutdown()
+  }
+
+  @Test
+  def evictedSessionWithLiveReferenceShouldNotBeClosedImmediately(): Unit = {
+    // create session
+    val session = connector.openSession()
+    val clientMock = session.unwrap()
+
+    // check that first session is still in cache (the same RiakClient instance will be returned)
+    val sameClientSession = connector.openSession()
+    assertSame(clientMock, sameClientSession.unwrap())
+    assertSame(session.unwrap(), sameClientSession.unwrap())
+    sameClientSession.close()
+
+    // evict session (second opening is required due to cache size == 2)
+    connector.openSession(Some(HostAndPort.hostsFromString("tmp1", 0).toSeq))
+    connector.openSession(Some(HostAndPort.hostsFromString("tmp2", 0).toSeq))
+
+    // check that first session was evicted and closed
+    assertNotSame(clientMock, connector.openSession().unwrap())
+    verify(clientMock, never()).shutdown()
+    assertFalse(session.isClosed)
+
+    session.close()
+    // check that session was closed
+    verify(clientMock, only()).shutdown()
+  }
+
+  /**
+    * This implicit conversion adds set() method to any object and allows to set any internal field including private.
+    * Such approach is used instead of usual reflection due to issues with receiving class of scala object
+    */
+  private implicit def reflector(obj: AnyRef) = new {
```

Comment:
Let's get rid of it by using:
 FieldUtils.writeField(RiakConnector, "com$basho$riak$spark$rdd$connector$RiakConnector$$sessionCache", sessionCacheSpy, true )